### PR TITLE
Scripts - collect dlls for Polyhedron demo with cygwin

### PR DESCRIPTION
**Pull request by janetournois**
## Summary of Changes

This PR introduces a new bash script for cygwin that collects all needed `.exe` and `.dll` to run the Polyhedron demo and its plugins.
It should be run from inside the demo build directory.
Maybe we should make it more general so that it can be used directly in the release creation process @lrineau @maxGimeno ?

## Release Management

* Affected package(s): Scripts

## TODO
- [x] Once the archive is OK, chancge the scripts for the doc that make the links to the demo

### CODE REVIEW

**Review by lrineau — APPROVED** [review]

**Comment by lrineau** [comment]
During the weekly GeometryFactory meeting on Monday, I have noted that we should:
  - test that scripts,
  - and then somehow integrate it in our automatic testsuite.

@maxGimeno Can you manage those tasks? I do no imply you do all of it by yourself, but I would like you to coordinate the effort.

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+declare target_directory="CGAL_demo_with_dlls"
+if [[ ! -d "$target_directory" ]]
+then
+  mkdir $target_directory
+fi
+
+copy_dll()
+{
+  local dll_full_path="`cygpath --unix --absolute $1`"
```
Comment (maxGimeno) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:15]:
```suggestion
  local dll_full_path=$(cygpath --unix --absolute "$1")
```

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+files=($config/*.exe)
+files+=($config/*.dll)
+files+=(Plugins/*/$config/*.dll)
+
+for file in "${files[@]}"; do
+
+  # copy exe or dll
+  copy_dll $file $target_directory
```
Comment (maxGimeno) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:28]:
```suggestion
  copy_dll "$file" "$target_directory"
```

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  # list and copy dependencies
+  cygcheck $file | while read -r dll ; do
+  
+    copy_dll $dll $target_directory
```
Comment (maxGimeno) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:33]:
```suggestion
    copy_dll "$dll" "$target_directory"
```

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+declare target_directory="CGAL_demo_with_dlls"
+if [[ ! -d "$target_directory" ]]
+then
+  mkdir $target_directory
+fi
+
+copy_dll()
+{
+  local dll_full_path="`cygpath --unix --absolute $1`"
+  echo "copy " $dll_full_path " to " $2
```
Comment (maxGimeno) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:16]:
```suggestion
  echo "copy $dll_full_path  to  $2"
```

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  cp $dll_full_path $2
```
Comment (maxGimeno) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:17]:
```suggestion
  cp "$dll_full_path" "$2"
```

Comment (lrineau) [reply Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:17]: @maxGimeno I suggest you commit the changes directly in Jane's branch. Anyway, you will test the script.

**Comment by maxGimeno** [comment]
qwindows.dll is still missing

**Comment by lrineau** [comment]
> qwindows.dll is still missing

That is the QPA plugin, loaded at runtime.

**Comment by maxGimeno** [comment]
Yes, which means it is not detected by Jane's script. We still need to collect it somehow.

**Comment by lrineau** [comment]
> Yes, which means it is not detected by Jane's script. We still need to collect it somehow.

We could add to the script:
```shell
cp "$(which qwindows.dll)" "$target_directory"
```

**Comment by janetournois** [comment]
actually it should be found in `platforms/qwindows.dll`.
do `which` can't find it like that

**Comment by janetournois** [comment]
We had worked on that in the issue #188

**Comment by lrineau** [comment]
We should probably use the tool https://doc.qt.io/qt-5/windows-deployment.html at first, so that Qt5 DLLs are tweaked as documented:
> In case Qt was built with the configure switch -relocatable turned off, windeployqt replaces the hardcoded local paths in Qt5Core.dll by relative ones.

And then use Cygwin `cygcheck` on the plugins, avoiding to overwrite `Qt5Core.dll`.

**Comment by maxGimeno** [comment]
Or we can simply define a Qt5_DIR in the script pointing to the right directory.

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 6]
```diff
@@ -29,9 +30,10 @@ for file in "${files[@]}"; do
...
   # list and copy dependencies
   cygcheck "$file" | while read -r dll ; do
-  
+
     copy_dll "$dll" "$target_directory"
-  
+
   done; #check dependencies
-    
+  mkdir -p "$target_directory/platforms"
```
Comment (janetournois) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:37]: this should happen outside of the for loop

**Comment by lrineau** [comment]
Only the tool `windeployqt` can "[replace] the hardcoded local paths in Qt5Core.dll by relative ones." I wonder if users of the precompiled demos will have to define `QT_QPA_PLATFORM_PLUGIN_PATH`, if we just copy the DLL files.

**Comment by maxGimeno** [comment]
We provide every dll in the directory, no one should have to modify anything. 
"In case Qt was built with the configure switch -relocatable turned off [..]" we have the hand on that, as everything comes from Gauguin. We just have to test it and we'll see.

**Comment by janetournois** [comment]
IIRC, 
- if Qt was installed with the installer, adding this dll is not needed
- otherwise (the installation directory of Qt was copied or moved on the same computer, e.g.)
-- if `platforms/qwindows.dll` is next to `Polyhedron_3.exe`, the system finds it itself
-- otherwise `QT_QPA_PLATFORM_PLUGIN_PATH` should de defined

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 7]
```diff
@@ -3,9 +3,13 @@
 #use this script from inside the build directory of the Polyhedron demo
 #Needs the Qt5_DIR env variable set to <Qt5_ROOT>/lib/cmake/Qt5
 
-declare config="Release"
+#No config : in autotest_cgal we use NMake as generator
+#If using MSVC Generator, declare config="Release"
+
+
+declare config=""
```
Comment (lrineau) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:10]: Then why not take config as an extra argument?
```suggestion
#No config : in autotest_cgal we use NMake as generator
#If using MSVC Generator, declare config="Release"


declare config="$2"
```

File: Scripts/developer_scripts/autotest_cgal [hunk 8]
```diff
@@ -655,8 +655,8 @@ echo 'COLLECTING RESULTS';
 ./collect_cgal_testresults_from_cmake;
 if [ -n "\$COLLECT_DEMOS_BINARIES" ]; then
   echo 'COLLECTING DEMOS BINARIES';
-  echo "cp ${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh ${CGAL_BINARY_DIR}/test"
-  cp "${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh" "${CGAL_BINARY_DIR}/test"
+  echo "cp ${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh ${CGAL_BINARY_DIR}/test \"\""
+  cp "${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh" "${CGAL_BINARY_DIR}/test \"\""
```
Comment (lrineau) [inline Scripts/developer_scripts/autotest_cgal:659]:
```suggestion
  echo "cp ${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh ${CGAL_BINARY_DIR}/test"
  cp "${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh" "${CGAL_BINARY_DIR}/test"
```

`cp` does not take three arguments.

File: Scripts/developer_scripts/autotest_cgal [hunk 9]
```diff
@@ -655,14 +655,14 @@ echo 'COLLECTING RESULTS';
...
-  cp "${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh" "${CGAL_BINARY_DIR}/test \"\""
+  echo "cp ${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh ${CGAL_BINARY_DIR}/test"
+  cp "${CGAL_TEST_DIR}/../developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh" "${CGAL_BINARY_DIR}/test"
 EOF
   cat >> "$file" <<'EOF'
   for demo_dir in *_Demo; do
     echo "pushd ${demo_dir}"
     pushd "${demo_dir}"
-    bash ../cgal_demo_copy_all_dlls_cygwin.sh "${demo_dir}_with_dlls"
+    bash ../cgal_demo_copy_all_dlls_cygwin.sh "${demo_dir}_with_dlls \"\""
```
Comment (lrineau) [inline Scripts/developer_scripts/autotest_cgal:665]: I do not understand. Should not it be the following instead?
```suggestion
    bash ../cgal_demo_copy_all_dlls_cygwin.sh "${demo_dir}_with_dlls" ""
```

Comment (maxGimeno) [reply Scripts/developer_scripts/autotest_cgal]: yes you are right

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-74.shtml

**Comment by lrineau** [comment]
@janetournois @maxGimeno This PR is tested but it still has a TODO at its first message (https://github.com/CGAL/cgal/pull/5430#issue-567750151). What should be done?

**Comment by maxGimeno** [comment]
the remaining todo is to update the paths to the demo in the doc. I don't remember where it is done

**Comment by maxGimeno** [comment]
@lrineau with my last commit, this PR should be all good. I guess the scripts in Maintenance are not used directly from the branch though, you'll have to update them locally from this branch.

**Comment by lrineau** [comment]
I have add "Under Testing" so that we can test the full scripts chain tomorrow.

**Comment by maxGimeno** [comment]
Success in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-89.shtml

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+then
+  mkdir $target_directory
+fi
+
+copy_dll()
+{
+  local dll_full_path=$(cygpath --unix --absolute "$1")
+  echo "copy $dll_full_path to $2"
+  cp "$dll_full_path" "$2"
+}
```
Comment (lrineau) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:23]: Maybe that function could ignore all DLLs that are in `C:\Windows`.

File: Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh [hunk 11]
```diff
@@ -19,7 +19,14 @@ copy_dll()
...
   echo "copy $dll_full_path to $2"
-  cp "$dll_full_path" "$2"
+  #necessary for paths with spaces in them, like "Visual Studio" or "Program Files", etc.
+  OLDIFS=$IFS;
+  IFS=$'\n';
+  #remove all dlls from system and Visual
+  if ! [[ "$dll_full_path" =~ "Visual" ]] && ! [[ "$dll_full_path" =~ "system32" ]]; then
+    cp "$dll_full_path" "$2"
+  fi
+  IFS=$OLDIFS
```
Comment (lrineau) [inline Scripts/developer_scripts/cgal_demo_copy_all_dlls_cygwin.sh:29]: Everything is quoted. I do not see why you modify `IFS`.

**Comment by lrineau** [comment]
@maxGimeno Please remerge this PR in `integration`, so that we can evaluate how it behaves.

**Comment by maxGimeno** [comment]
successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-98.shtml

**Comment by maxGimeno** [comment]
I tested the resulting tar on Blake and it worked. The new size is around 228 Mo I

**Comment by maxGimeno** [comment]
Success in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-109.shtml

**Pull request merged**